### PR TITLE
fix deadlock of update shards feature

### PR DESCRIPTION
**Pull request by mjneil**
```
$ go test ./... -run=TestUpdateShards
2023/02/10 18:45:15 flushing records reason=batch size,  records=7
2023/02/10 18:45:15 flushing records reason=flush interval,  records=1
2023/02/10 18:45:20 flushing records reason=batch size,  records=7
2023/02/10 18:45:20 flushing records reason=flush interval,  records=1
2023/02/10 18:45:20 flushing records reason=flush interval,  records=5
--- FAIL: TestUpdateShards (110.02s)
    --- FAIL: TestUpdateShards/scale_up_8_->_32_shards (100.00s)
        require.go:794: 
            	Error Trace:	producer_test.go:559
            	            				asm_amd64.s:1594
            	Error:      	Received unexpected error:
            	            	rate: Wait(n=1) would exceed context deadline
            	Test:       	TestUpdateShards/scale_up_8_->_32_shards
        require.go:301: 
            	Error Trace:	producer_test.go:572
            	Error:      	test timeout
            	Test:       	TestUpdateShards/scale_up_8_->_32_shards
FAIL
FAIL	github.com/mjneil/kinesis-producer	110.266s
ok  	github.com/mjneil/kinesis-producer/deaggregation	(cached) [no tests to run]
?   	github.com/mjneil/kinesis-producer/loggers/kplogrus	[no test files]
?   	github.com/mjneil/kinesis-producer/loggers/kpzap	[no test files]
?   	github.com/mjneil/kinesis-producer/main	[no test files]
?   	github.com/mjneil/kinesis-producer/pb	[no test files]
FAIL
```

### CODE REVIEW

**Comment by mjneil** [comment]
```
$ go test -run=TestUpdateShards
2023/02/13 17:34:37 flushing records reason=batch size,  records=7
2023/02/13 17:34:37 flushing records reason=flush interval,  records=1
2023/02/13 17:34:41 flushing records reason=flush interval,  records=8
2023/02/13 17:34:42 flushing records reason=flush interval,  records=8
2023/02/13 17:34:47 flushing records reason=batch size,  records=31
2023/02/13 17:34:47 flushing records reason=flush interval,  records=1
2023/02/13 17:34:52 flushing records reason=flush interval,  records=1
2023/02/13 17:34:52 flushing records reason=batch size,  records=31
2023/02/13 17:34:52 flushing records reason=flush interval,  records=30
PASS
ok  	github.com/mjneil/kinesis-producer	20.624s
```

File: main/main.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+	if l := os.Getenv("AWS_DEBUG_LOGGING"); l != "" {
+		conf = conf.WithLogLevel(aws.LogDebug)
+	}
+
+	return kinesis.New(session.Must(session.NewSession(conf)))
+}
+
+func main() {
+	client := newKinesisClient("eu-central-1")
+	shards, _, _ := producer.GetKinesisShardsFunc(client, "prod-ssai-ad-tracking-4772533039001")(nil)
```
Comment (bcasinclair) [inline main/main.go]: Should that stream name be in there?

Comment (mjneil) [reply main/main.go]: oops 🤦

**Comment by mjneil** [comment]
@petethepig I noticed you've been pulling in changes from my fork, you will want this critical bug fix if you're using the shard map feature

**Comment by petethepig** [comment]
@mjneil Thanks for the heads up!

**Review by bcasinclair — APPROVED** [review]
Not sure I fully understand the implications of the idle connections changes as my low level kinesis knowledge is limited but the test case looks good

**Review by ferpart — APPROVED** [review]

**Pull request merged**